### ISSUE TITLE
Screensaver keeps freezing

### ISSUE THREAD

**Issue by apbitner** [AerialScreensaver/Aerial#31]
Been using Aerial for months but lately I'll come back to my computer to find the screen frozen on a frame from an Aerial movie. I also updated to Monterey recently so it could possibly be related to that 🤷🏻‍♂️

macOS Monterey 12.0.1
Mac mini (M1, 2020) - 8gb
Aerial Companion 1.2, Screensaver 2.3.3

**Comment by glouel** [comment]
Hey there !

Hmm, frozen on a frame is a bit of a new one. This used to happen a lot with streaming but unless you were creative with your cache settings Aerial should rarely if ever stream nowadays. 

I would very much like to see a log if you have one in any case. How frequently does that happen ?

**Comment by apbitner** [comment]
Sure, here's my Aerial log, hope this is what you wanted. I'm not sure if it happens every time Aerial runs but it seems to happen quite often these days.

[AerialLog.txt](https://github.com/glouel/AerialCompanion/files/7551065/AerialLog.txt)

**Comment by glouel** [comment]
It is thanks ! I'll have a look tomorrow and get back to you. 

Sorry for the issue in any case, I did notice some weirdness during the Monterey beta process (which were mostly fixed) but that's a new one.

**Comment by glouel** [comment]
@apbitner so I checked your log and have a hard time finding any issue. Next time this happens, can you note your exact local time before you wake up, and give me a log with that timestamp so I can try digging a bit more ? It doesn't look like you were streaming in any case so it's something new. 

Also I didn't ask but a couple of things : 
- When the screen is frozen, can you still wake up your computer normally ? 
- I suppose you would have catched it but it doesn't happen on the same video I assume ?

Thanks !

**Comment by apbitner** [comment]
Hello,

Actually it just happened before I filed this bug last evening, but I will
take note of it next time.

I am able to wake the computer when I move my mouse and it happens on
different videos.

**Comment by glouel** [comment]
Thanks for your quick answer, that's super helpful and I maybe have a couple of ideas on what's happenning, but you may be able to clarify the timeline and your Mac settings a bit.

Here's what I'm seeing : 

- at 17:34:50 Aerial kicked in and was launched as a ScreenSaver
- at 18:30:39 it started playing a video of the ISS, Mexico to NY which is about 5:30 long. I'm expecting it to stop at 18:36:09
- **at 18:35:00 macOS is telling me that the screensaver will be killed**. I believe it gets killed then but may still be running idle. 
- at 19:43:04 you launched Aerial's control panel in setup

So here's my thoughts, can you clarify that you actually woke up your computer around 19:43, and not at 18:35 ? 

The thing is, it'a about exactly 1h between Aerial starts and the "freeze", which is a bit peculiar. I looked up a couple of previous willStop event (that's when macOS tells me that the screensaver will be killed) : 

Nov 14: 
- started at 19:50:42
- willStop at 20:50:51
Nov 15: 
- started at 19:10:03 
- willStop at 20:10:12 

So I'm thinking you have some combination of macOS settings that makes macOS sends me a screensaver will stop event, yet macOS doesn't kill Aerial. 

First, can you tell me about your setting here : 

<img width="893" alt="Capture d’écran 2021-11-17 à 15 31 05" src="https://user-images.githubusercontent.com/37544189/142219836-8b4794bf-c6fd-443d-aa6d-8635e32dd417.png">

Then can you show me your Energy Settings in macOS ? I'm wondering if you are not putting your hard drive (or maybe screen and that fails) to sleep after one hour, which would cause macOS to tell me to stop the saver. Here's the panel on my iMac, you will have some extra settings on a laptop which may be what's causing this. 

<img width="893" alt="Capture d’écran 2021-11-17 à 15 27 30" src="https://user-images.githubusercontent.com/37544189/142219902-d416592b-f063-4406-8cd2-0b6d2a1e91c7.png">

I'm pretty confident that the answer is around here. As to why I'm told by macOS to stop the screensaver, I don't know, but when I get the event I pause the video, restore the brightness settings and get ready to be killed (which macOS does, I don't have control over that). 

It's possible macOS is sending me the event yet doesn't plan to kill Aerial, I did introduce that pause fairly recently to make Macs more reactive when you wake up (it can take 6/7 seconds for macOS to kill Aerial, with the video still playing in the background, while you are on your desktop already). It's possible that there's a mess around here with events that I need to clear up. 

Again, thanks for following up and let me know what you can about your settings, pretty sure we'll find a solution around here.

**Comment by apbitner** [comment]
Hello,

I think you're on to something. The screensaver comes on after 5 minutes,
and here are my energy settings:

[image: image.png]
I do prevent my Mac Mini from going to sleep because I had been having
issues with my Bluetooth mouse/keyboard not reconnecting after it went to
sleep, and enabling that checkbox seems to have fixed the issue. It's weird
though because I downloaded Aerial months ago and only recently noticed the
freezing issue, and as I mentioned the only significant change recently has
been the upgrade to Monterey.

**Comment by glouel** [comment]
Can you attach your screenshot on github ? Attachements doesn't work on mail. Thanks !

**Comment by apbitner** [comment]
Ah, here you go.

![image](https://user-images.githubusercontent.com/22176302/142234856-34c38945-4615-4b70-a306-539946bcd2c8.png)

**Comment by glouel** [comment]
Ohhhhhh... 

Ok so I'm a bit confused now, can you clarify if your screen *actually* goes to sleep with those settings ? Is the screen asleep and when you get back you wake it up and get a paused Aerial? I'm kinda thinking your screen never goes to sleep here, am I getting this right ? 

I'm wondering whether I'm missing an event to restart the screensaver or if it's something else. In any case, this is definitely this. 

Just to be clear, what happens with my settings is (actually, I'm only 95% sure) that macOS kill Aerial before putting my screen to sleep. But maybe I'm getting all this wrong. Anyway it should be easy to fix.

(And it's quite possible the bug is newly introduced in Monterey also)

**Comment by apbitner** [comment]
In the past, the screensaver would come on after 5 minutes, and then the display would go to sleep after an hour. Now, Aerial seems to freeze after an hour and the display no longer goes to sleep. 

I will do another test this evening, but yes the issue seems to be that before Monterey my display would go to sleep after 1 hr. And now after 1 hr the screensaver freezes and the display never goes to sleep.

**Comment by glouel** [comment]
Ok that makes perfect sense now, thanks ! To be clear I don't think it's "frozen" as in crashed, it's just that I pause the video waiting to be killed, which apparently never comes. The screen saver itself isn't in charge of putting your screen to sleep and macOS is in charge of killing the screensaver too, so could be some newish bug with those settings in Monterey. 

I'm running 12.1 beta here but I'll try to replicate your settings tonight to see what happens and get back to you.

**Comment by glouel** [comment]
Very quick update, my first test with your settings (1h screen off, and never put computer to sleep), screen went to sleep fine after 1 hour. 

I'll keep those settings for a few days to see if I can trigger your issue but this may be a M1/12.0 bug with macOS.

**Comment by glouel** [comment]
This may be unrelated, but found this post : https://old.reddit.com/r/MacOS/comments/qj9cn4/macbook_wont_sleep_on_monterey/

Maybe try disabling power nap to see if it has any effect ?

**Comment by apbitner** [comment]
I left my computer for 2 hours today and came back and display was asleep, so will keep an eye on it. I'm also on a Mac mini, not a Macbook, so I don't have power nap.

**Comment by apbitner** [comment]
Spoke too soon, happened again. I stepped away from my computer around 7:30pm, came back around 9:08pm and the screensaver was paused on the screen. FYI, I changed my display to go to sleep after 15 minutes instead of an hour in hopes of making this easier to test.

[AerialLog.txt](https://github.com/glouel/AerialCompanion/files/7567533/AerialLog.txt)

**Comment by glouel** [comment]
A few thoughts 
- first it looks like they may have removed Power Nap from M1s (I definitely have it on my Intel iMac, it's not notebook only).
- second, did you check the reddit thread ? This is all about issues with bluetooth too, specific to 12.0. What they describe is they see the screen fading to sleep and waking up instantly then never going again. This may be what's happening to you too.  
- sometimes `pmset -g` may give you more info about what blocked it 

I've been running your settings here for a couple of days without an issue so I'm fairly confident this is a macOS thing. 

If pmset doesn't give you a clue, I would probably try disabling again the "Prevent going to sleep" and see if that helps or not. 

You could also try another screensaver for a couple of days to see if you have the same issue, but I believe you will at this point. 

This may not be incredibly helpful but that's all I can think of as of now, hopefully whatever that bug is is fixed in 12.1 for you. Bluetooth is the root of so many issues, I hoped it would get better on Apple Silicon but apparently it didn't 😩

**Comment by TheRealChannelD** [comment]
Hi there, I wanted to see if this is still being researched.

I seem to have just started having this issue with 12.4/12.5 of Mac Monterey.   I did go to the latest beta 1.2.2 (3.1.0beta3) and that didn't help.

For me, I have an external monitor and my Mac open (unclamshelled).   The lockup happens between 12-24 hours.   The only screen that ever locks up is the main Mac.   The external 34" monitor never has the issue.  In fact after the main Mac lockups, I have let the external monitor go for another 12+ hours with no issues.  Both are using different screensavers if that's helpful information.

Closing the laptop and reopening it fixes it, but it will eventually lockup again (usually during a conference call when I am using both screens).

Let me know if there are debug logs or something I can attach.  I am using a Intel MacBookPro16,2 32GB RAM.

Thanks!
